### ISSUE TITLE
New feature: MapTileApproximater with several sources

### ISSUE THREAD

**Issue by monsieurtanuki** [osmdroid/osmdroid#646]
As suggested by @secaw in #640:

> Regarding MapTileApproximater and offline sources: users may be able to put custom offline maps on top of online sources. So what happens when being offline is that if e.g. you have an offline map from zoom 6 to 15 and are zooming into 16 you then get a scaled tile from cached online source of zoom level 5 instead of scaled offline tile from level 15, which would not be the desired outcome. I hope this description makes sense? So I think including all available sources for approximation would be very desirable. Thanks a lot for thinking about looking into this! Can imagine that it's not that trivial.

What about putting several `MapTileApproximater`s in your `MapTileProviderArray`? No, I don't think it would work because even if the good tile will eventually be computed, it will be overwritten by the next approximater.

What about having several `IFilesystemCache`s in the constructor? Then the question will be: to approximate level 16, am I better off with level 5 of my favorite source or with level 15 of my default source? And how can I design a simple set up for that?

Another idea would be to have the memory cache as the only input. But we cannot be sure it will be fully populated.

**Comment by secaw** [comment]
Just a quick thought without having the code in front of me: What about setting up a method for each tileProvider, something like `int getMaxAvailableZoomLevelForTileApproximation` and then pass an `IFilesystemCache` Array to `MapTileApproximater` like you suggested, which will approximate from the highest available zoom level compared from all providers including system cache?

**Comment by monsieurtanuki** [comment]
For the moment we don't have providers (extending `MapTileModuleProviderBase`) as input but `IFileSystemCache`s. The former use callbacks and write to the memory cache, the latter give direct access to the Drawable through method `Drawable loadTile`.
And `IFileSystemCache`s don't even have a min/max zoom!

What I suggest in order to implement both your features (several sources + zoom relevancy zone) would be:
* to create in `MapTileApproximater` method `void addSource(ITileSource tileSource, IFileSystemCache reader, int minZoom, int maxZoom)`
* to create a version of method `approximateTileFromLowerZoom` that take additional parameters such as tileSource, reader, minZoom and maxZoom
* to remove the `IFileSystemCache` parameter from the constructor and to ignore `setTileSource` statement in order to remove confusion - the only way to add a source will be to use method `addSource`

What do you think?

**Comment by secaw** [comment]
Sorry, my mistake. I was thinking about including the locally available file archives (not `SystemCache`s) in `MapTileApproximater`. So it's not about different sources within the system cache. If you have a tile for the region from a file archive (any source) with zoom level 15 and are zooming into 16, you probably would want this tile to be scaled instead of the cached online tile from level 5.
Hope to have been a bit clearer now...

**Comment by monsieurtanuki** [comment]
I mentioned the `ITileSource` because it's used by `IFileSystemCache` in order the get the tile bitmap.
But you're right, local file archives don't have the concept of `ITileSource`.

What we need is:
* to add a simple method `public Drawable loadTile(MapTile pMapTile)` to `MapTileModuleProviderBase.TileLoader`, that will be a simple rewriting of the existing `protected Drawable loadTile(MapTileRequestState pState)` (and may even replace it altogether)
* to create in `MapTileApproximater` method `void addProvider(MapTileModuleProviderBase provider, int minZoom, int maxZoom)`
* to create a version of method `approximateTileFromLowerZoom` that take additional parameters such as provider, minZoom and maxZoom
* to remove the `IFileSystemCache` parameter from the constructor and to ignore `setTileSource` statement in order to remove confusion - the only way to add a source will be to use method `addProvider`

It looks better now, doesn't it?

**Comment by secaw** [comment]
Yes, sounds better!
But how do we check, which tile is the closest (max available zoom) from which provider, that should be used.

Say you are offline and zoom into level 16 and you have these tiles available for the region:
- From offline map file 1 with zoom level 15
- From offline map file 2: none for this region
- From CacheProvider with zoom level 5

The tile from offline map file 1 should be taken for approximation, because it's the closest (the one where we actually come from when zooming in).
That's why I proposed the method in my first comment, so we can check the 'maxAvailableZoomLevelForTileApproximation' for each provider for the specific region before getting the tile. This would avoid expensive multiple tile loading before displaying the right one. At least that's how I imagine it now. Unfortunately I'm more less offline at the moment, so can't check the code details..

**Comment by monsieurtanuki** [comment]
It's not 100% clear in my mind; I would need more details about method `int getMaxAvailableZoomLevelForTileApproximation` (to be created). Could you confirm that:
* it should be created in class `MapTileModuleProviderBase`
* it has no parameter XOR it has one parameter `MapTile` XOR something else

My first assumption was that you wanted this method in order to say: "this provider GENERALLY gives ugly bitmaps beyond zoom 15, so it should be used only between 0 and 15 for `MapTileApproximater`". But from your latest message you seem to want it with one parameter (something like a `MapTile` I guess) for optimization reasons.
I'm a big fan of optimization, but I don't think we need this new method here.

The algorithm would be something like:
* I create a `MapTileApproximater` and add (ordered) providers 1, 2 and 3 to it
* I'm offline and I don't have a bitmap for this tile
* I call a non static version of `MapTileApproximater.approximateTileFromLowerZoom(MapTile mapTile)` (to be created)
* for `mapTile.getZoom() - 1`, I try provider 1. If it doesn't work I try provider 2. If it doesn't work I try provider 3
* If it doesn't work it try again for `mapTile.getZoom() - 2`, and so on

Wouldn't it work exactly the way you want it to? And we don't even need the new max zoom method.

Tell me if we agree on the goal:
* being able to add several providers (`MapTileModuleProviderBase`) to `MapTileApproximater`
* when `MapTileApproximater.getTileLoader().loadTile(MapTile mapTile)` is called, we return an approximation from the best possible bitmap, which comes from the best possible zoom level among all providers, and from the first provider on the ordered providers list for that best zoom level

**Comment by secaw** [comment]
OK, very clear, got you. I absolutely agree on the goal, exactly what I thought. If we don't need the proposed method (yes, with parameter MapTile) for optimisation, great! We don't even need parameters `minZoom` and `maxZoom` for new method `addProvider`, do we (might come from misunderstood first assumption you are talking about)?

**Comment by monsieurtanuki** [comment]
Exactly!
I'll start the dev as soon as my pending PR pass.

**Comment by secaw** [comment]
Fantastic, looking forward!

**Comment by spyhunter99** [comment]
this should be an opt in feature. Multiple offline sources could be normal for certain applications (road maps and sat maps for example), auto switching tiles source could be problematic

**Comment by monsieurtanuki** [comment]
@spyhunter99 The idea is to have explicit code in `MapTileProviderBasic`'s constructor. Something like:
`approximationProvider.addProvider(cachedProvider);
approximationProvider.addProvider(archiveProvider);`

> this should be an opt in feature

"Basic" users will use `MapTileProviderBasic` (with the feature on), and more advanced users can extend `MapTileProviderArray` and customize their use of `MapTileApproximater.addProvider`. For a more dynamic use I can imagine a `MapTileApproximater.clearProviders` method.
But I'm not sure I've fully covered your comment. Have I?

**Comment by secaw** [comment]
@spyhunter99 I think this is a misunderstanding from the title 'MapTileApproximater with several sources' which should more accurately be something like 'Include several providers in MapTileApproximater'.

Before `MapTileApproximater` - when offline - the zoom in behavior with no available tile was that the current view with tiles from a file archive was just scaled as much as you zoomed in. With the current implementation of `MapTileApproximater` this does not happen anymore, instead you get a much larger scaled version of your cached online tile source from many levels below (because `MapTileApproximater` currently only uses the cache provider).

The proposal from my perspective basically is simply to restore the behavior before the current implementation of `MapTileApproximater` in order to be able to zoom into the current view (including map archive tiles) with the huge improvement of `MapTileApproximater` that you are actually able to pan around this scaled view!

**Comment by secaw** [comment]
Finally was able to test the new `MapTileApproximater` with archive providers: PERFECT !!! It's now seamlessly behaving like any user would expect when zooming into and panning around the map no matter which part of the map is coming from which provider. Thanks.

**Comment by spyhunter99** [comment]
i'll second that, probably the best contribution to osmdroid this year

**Comment by monsieurtanuki** [comment]
@secaw @spyhunter99 thank you!
Now I've switched to #329. I can't see anything else to do on the raster tile display part of osmdroid.

**Comment by secaw** [comment]
@monsieurtanuki 
In my logcat I noticed some nullpointer exceptions coming from `MapTileApproximater` when closing the map or rebuilding the provider array. Digging deeper, I saw they were coming from SQLite archives as well as the `MapTileSqlCacheProvider` within `MapTileApproximater`, because I used the same instances for both `MapTileApproximater` and `MapTileProviderArray`. The problem seems to be that the SQLite instances get closed by `MapTileProviderArray` but `MapTileApproximater` is still trying to connect to them.
Can you reproduce this, since you are also using the same instances of `cacheProvider` (`MapTileSqlCacheProvider`) for both `MapTileApproximater` and `mTileProviderList` in `MapTileProviderBasic`?
As a work around I have created separate instances for each provider etc. to pass on to `MapTileProviderArray` and `MapTileApproximater` which fixes the null pointer issue, but it seems that there's a lot more calculation going on now, the map starting to flicker especially at startup and pretty much lagging overall when offline. Do you maybe know a better solution?

**Comment by monsieurtanuki** [comment]
@secaw I won't be able to work on this subject in the following days and I don't have access to the code right now, but my first reactions are:
* the standard use is to reuse the providers in `MapTileApproximater` (which provokes NPE) and not to have additional providers (as in your work around)
* part of the solution may be to "detach" providers in reverse order
* part of the solution may be to create a new method in `MapTileApproximater` called `clearProviders` that would clear its providers' list

**Comment by secaw** [comment]
@monsieurtanuki 
Thanks for your feedback.
According to your suggestions I did some try and error myself:

1.) I tried a new detach override for `MapTileApproximater` like this:
```
  @Override
  public void detach() {
    mProviders.clear();
    super.detach();
  }
```
which at least limited the number of SQLite NPEs and file closed exceptions.

2.) I also tried to revert the detach order in `MapTileProviderArray` like this:
```
public void detach() {
  synchronized (mTileProviderList) {
    /*
    // OLD
    for (final MapTileModuleProviderBase tileProvider : mTileProviderList) {
      tileProvider.detach();
    }
    */
    // NEW
    for(int i = mTileProviderList.size()-1 ; i >= 0; i--){
      mTileProviderList.get(i).detach();
    }
  }
  ...
}
```
but still get SQLite NPEs and file closed exceptions (not all anymore from `MapTileApproximater`).

I'm not really able to dig any deeper, but maybe that helps..

**Comment by monsieurtanuki** [comment]
@secaw Just tried (with `StarterMapFragment`) but could not reproduce the NPE.
Had a look at the code: wouldn't it make more sense to detach the tile writer AFTER the providers have been detached?
In `MapTileProviderBasic` that would look like this:
```
public void detach(){
    super.detach();
    if (tileWriter!=null)
        tileWriter.onDetach();
    tileWriter=null;
}
```
Could you try that? (as I cannot reproduce the bug)

**Comment by secaw** [comment]
@monsieurtanuki 
Alright, the source of the problem was actually something different! Since I did not use `MapTileProviderBasic`, but `MapTileProviderArray` directly, tile loading was stuck in an infinite loop with expired tiles and when offline, because 'isDowngradedMode()' always returned false. I created my own `MapTileProviderArray` now based on `MapTileProviderBasic`, which solved this problem and actually a few others. I can imagine though that the `MapTileApproximater.detach()` method would be useful in similar situations like this and generally a clean way to terminate it:
```
@Override
public void detach() {
  super.detach(); // first detach super
  mProviders.clear();
}
```

Also I changed 'isDowngradedMode()' in `MapTileProviderArray` from
```
return mNetworkAvailabilityCheck != null && !mNetworkAvailabilityCheck.getNetworkAvailable();
```
to
```
return (mNetworkAvailabilityCheck != null && !mNetworkAvailabilityCheck.getNetworkAvailable())
    || !useDataConnection();
```
because you also want to keep expired tiles when the provider is 'forced' offline through `setUseDataConnection()`. I think this also should be taken over in osmdroid master. Do you agree?

Furthermore I managed to improve performance in `MapTileProviderArray` for my own implementation, but I'm not sure, if those changes can be generalized.

All is running extremely smooth now. Again - a fantastic overall improvement compared to when I first started using osmdroid beginning of this year.

**Comment by monsieurtanuki** [comment]
@secaw I'm glad that you fixed the issue on your app and that you enjoy a smooth tile display.

I agree with your suggestions on `MapTileApproximater.detach()` and on `MapTileProviderBasic.isDowngradedMode()`. I do believe though that ``MapTileProviderArray.isDowngradedMode()` should remain `false` by default, it's just an array after all, everything depends on its items, which may have nothing to do with internet connections.

Could you describe the other improvements you developed, or is that really too specific to your app?

**Comment by secaw** [comment]
@monsieurtanuki Yes I was also thinking about modified `isDowngradedMode()` in `MapTileProviderBasic` only.

Regarding the other changes I was talking about:
To begin with, I have just a slightly different provider concept compared to `MapTileProviderBasic`. From the start with osmdroid I had many problems with the fact, that in `MapTileProviderBasic` the `cacheProvider` was rendered in front of the `archiveProvider`, which in many situations led to a patchwork map display when offline with missing or even expired tiles. This may quite work as expected, if you have a consistent tile source, but the other part of my concept is, that I want to have source unaware archives (see #688). This is in order to make it most easy for users to just drop any archive file (e.g. from MOBAC) into the osmdroid folder and that will always be displayed, no matter what tile source selected (in my app users may then deactivate / activate any found archive file on the fly).
That's why I created my own `MapTileProviderArray` with a source unaware `archiveProvider` first and `cacheProvider` second, followed by `downloaderProvider` and now `approximationProvider`.

The further improvements in my `MapTileProviderArray` I was talking about are

(1) In `findNextAppropriateProvider`, I don't want to use an online provider not only if it should not use a data connection, but also if there is no network available:

OLD
```
providerCantGetDataConnection = !useDataConnection() && provider.getUsesDataConnection();
```
NEW
```
providerCantGetDataConnection = isDowngradedMode() && provider.getUsesDataConnection();
```
with `isDowngradedMode()` using the modified above version. Of course this change would not work with current `MapTileProviderArray`, as it doesn't have a `NetworkAvailabilityCheck` etc. as you state above.


(2) In `mapTileRequestExpiredTile`, I generally don't want to continue to look within an offline provider for a tile that is expired:

OLD
```
final MapTileModuleProviderBase nextProvider = findNextAppropriateProvider(aState);
```
NEW
```
MapTileModuleProviderBase nextProvider;
// We don't want to continue with an offline provider, if an available cached tile is just expired
boolean providerIsAnOfflineProvider = false;
if(ExpirableBitmapDrawable.getState(aDrawable) == ExpirableBitmapDrawable.EXPIRED) {
  do {
    nextProvider = findNextAppropriateProvider(aState);
    if (nextProvider != null) {
      providerIsAnOfflineProvider = !nextProvider.getUsesDataConnection();
    }
  } while ((nextProvider != null) && providerIsAnOfflineProvider);
} else {
  nextProvider = findNextAppropriateProvider(aState);
}
```
This might actually make sense to be taken over for osmdroid `MapTileProviderArray`, but I'm not sure, if there are use cases where expired tiles should be replaced by offline tiles? This wouldn't work anymore.
In my use case in the end it's just skipping another senseless tile loading round, but I like those little optimizations...
